### DISCUSSION TITLE
How Is HNSWPQ Implemented in Code?

### DISCUSSION THREAD

**Discussion by zzyuewo** [facebookresearch/faiss#3797] [General]
i am new to coding，In IndexHNSW.cpp, the storage is always  IndexPQ or not?
in void IndexHNSW::add() 
{  ....
   storage->add(n, x);
   ....
}
will it use the add method in IndexFlatCodes? because i see there is no add() in indexPQ and IndexPQ : IndexFlatCodes

in void hnsw_add_vertices()
{    ....
    std::unique_ptr<DistanceComputer> dis(storage_distance_computer(index_hnsw.storage));
     ....
}
but i think that DistanceComputer doesnt work for PQ

Which method is actually used by HNSWPQ? For these functions,(add, storage_distance_computer, set_query, add_with_locks)
i want to know how faiss combined HNSW and PQ in code, but i am confused

**Comment by rafayaar** [comment]
@zzyuewo 

- If `IndexHNSW` is combined with PQ, storage would typically be an instance of `IndexPQ`, which in turn inherits from `IndexFlatCodes`.
- The `add()` method in IndexHNSW would delegate the call to `IndexPQ::add()`, but since `IndexPQ` does not have its own add() method, it inherits the add() method from `IndexFlatCodes`. 
In this method, the vectors are PQ-encoded before being added to the index.

For the functions (add, storage_distance_computer, set_query, add_with_locks)

- `add()`: Adds PQ-encoded vectors to the HNSW graph.
- `storage_distance_computer()`: Returns a PQ-specific distance computer for searching.
- `set_query()`: Sets the query vector, likely compressed using PQ before searching.
- `add_with_locks()`: Adds vectors to the HNSW graph while ensuring thread safety, encoding them with PQ if used.

> i am new to coding，In IndexHNSW.cpp, the storage is always IndexPQ or not? in void IndexHNSW::add() { .... storage->add(n, x); .... } will it use the add method in IndexFlatCodes? because i see there is no add() in indexPQ and IndexPQ : IndexFlatCodes
> 
> in void hnsw_add_vertices() { .... std::unique_ptr dis(storage_distance_computer(index_hnsw.storage)); .... } but i think that DistanceComputer doesnt work for PQ
> 
> Which method is actually used by HNSWPQ? For these functions,(add, storage_distance_computer, set_query, add_with_locks) i want to know how faiss combined HNSW and PQ in code, but i am confused
